### PR TITLE
Add `\#` escaping

### PR DESCRIPTION
**Pull request by killercup**
Thanks for your suggestions in #88, @BurntSushi!

I added some tests, but I'm not sure they're covering all edge cases.

### CODE REVIEW

File: regex-syntax/src/parser.rs [hunk 1]
```diff
@@ -1047,19 +1058,19 @@ fn rev_concat(mut exprs: Vec<Expr>) -> Expr {
     }
 }
 
-// Returns ture iff the given character is allowed in a capture name.
+// Returns true if the given character is allowed in a capture name.
```
Comment (BurntSushi) [inline regex-syntax/src/parser.rs]: Hah. The "iff" actually isn't a typo. It means, "if and only if." It's leftover shorthand from getting a math degree. :P

File: regex-syntax/src/parser.rs [hunk 2]
```diff
@@ -801,14 +801,25 @@ impl<'a> Chars<'a> {
...
     }
 
+    /// Skip line comments and whitespace
     fn skip(&mut self) {
         if !self.ignore_space { return; }
         while self.cur < self.chars.len() {
+            // Handle escaping of `#`, i.e. don't start a comment with `\#`.
+            if !self.in_comment && (self.c() == '\\')
+                && ((self.cur + 1) < self.chars.len())
+                && (self.chars[self.cur + 1] == '#')
```
Comment (BurntSushi) [inline regex-syntax/src/parser.rs]: Arg. Get those redundant parens off my lawn!

File: regex-syntax/src/parser.rs (hunk 1 above)
Comment (killercup) [inline regex-syntax/src/parser.rs]: Oh, haven't seen that before. (Haven't read any math papers in English, though.) I'll change it to "if and only if".

File: regex-syntax/src/parser.rs [hunk 3]
```diff
@@ -801,14 +801,25 @@ impl<'a> Chars<'a> {
         }
     }
 
+    /// Skip line comments and whitespace
     fn skip(&mut self) {
         if !self.ignore_space { return; }
         while self.cur < self.chars.len() {
+            // Handle escaping of `#`, i.e. don't start a comment with `\#`.
+            if !self.in_comment && (self.c() == '\\')
+                && ((self.cur + 1) < self.chars.len())
```
Comment (BurntSushi) [inline regex-syntax/src/parser.rs]: Could you please use `checkadd(self.cur, 1)` here (and below)? I'm trying to be a bit paranoid about overflow in the parser. (Although I don't have any tests for it...)

File: regex-syntax/src/parser.rs (hunk 1 above)
Comment (BurntSushi) [inline regex-syntax/src/parser.rs]: Yeah, I can't recall it showing up much in papers. It's more like "I'm hand writing a proof and I'm going to gratuitously abuse shorthand."

File: regex-syntax/src/parser.rs (hunk 2 above)
Comment (killercup) [inline regex-syntax/src/parser.rs]: Your have your overflow paranoia, I have my parens. ;)

How much can I get rid of? Does `+` bind stronger than `<` which binds stronger than `&&`?

Comment (BurntSushi) [inline regex-syntax/src/parser.rs]: Yup. Exactly.

Comment (BurntSushi) [inline regex-syntax/src/parser.rs]: I'm usually not one to rely on memorization of precedence, but in these cases, I think if the precedence was anything else, you'd end up with a compile error.

**Comment by killercup** [comment]
@BurntSushi, I've added a commit using more checked adds and less parens.

Btw, there are a lot of small functions in there. I hope they all get inlined.

**Comment by BurntSushi** [comment]
Excellent! Thank you!

> Btw, there are a lot of small functions in there. I hope they all get inlined.

Yeah, not so concerned. There are other things I'm more concerned about affecting performance (like simplifying the AST and canonicalizing character classes). For another day!

**Comment by BurntSushi** [comment]
OK, crates.io updated! (`regex-syntax 0.1.1`)
